// forum/Posts.js

var Post = require("./Post"),
	sqlUtil = require("./sqlUtil");

var Posts = module.exports = exports = function(db, filter){
	this.db = db;
	this.filter = filter;
	
	this._posts = [];
	this.totalCount = null;
};

var fnData = function(){
		var results = [];
		this._posts.forEach(function(post){
			results.push(post.data());
		});
		return results;
	},
	
	fnFromSql = function(rows){
		var self = this;
		rows.forEach(function(row){
			var post = new Post();
			post.fromSql(row);
			self._posts.push(post);
		});
	},
	
	fnFields = function(){
		var fields = [],
			p = Post.prototype
		for (var key in p){
			if (typeof p[key] !== "function" && key.charAt(0) !== '_'){
				fields.push(key.toUnderScore());
			}
		}
		return fields;
	},
	
	fnLoad = function(range, callback){
		if(this.db){
			var self = this;
			this.db.all(sqlUtil.getSqlSelect(this, range), function(err, rows){
				if(err) throw err;
				self.fromSql(rows);
				if(callback){
					callback(self);
				}
			});
		}
	},
	
	fnTotalCount = function(callback){
		if(this.db){
			var self = this;
			this.db.all(sqlUtil.getSqlCount(this), function(err, rows){
				if(err) throw err;
				console.log(rows);
				if(callback){
					callback();
				}
			});
		}
	};

Posts.prototype = {
	_table: "posts",
	_id: "id",
	_posts: [],
	db: null,
	totalCount: fnTotalCount,
	
	fromSql: fnFromSql,
	fields: fnFields,
	load: fnLoad,
	data: fnData
};
